### ISSUE TITLE
Product includes P2 cache

### ISSUE THREAD

**Issue by basilevs** [eclipse-tycho/tycho#2510]
When building a product that contains a feature with root files, root files (in a compressed form) are left in 
`<install_dir>/p2/org.eclipse.equinox.p2.core/cache`

The resulting product distribution artifact is larger than it ought to be as root files are distributed twice - in an installed and in cached form.

### Steps to reproduce
- Unzip [demo.zip](https://github.com/eclipse-tycho/tycho/files/11746080/demo.zip)
- Run `bash org.basilevs.rootfilesdemo/populate_large_file.sh`
- Run `mvn clean package`
- Run `find org.basilevs.product/target/products/product/linux/gtk/x86_64 -size +900M  -ls`

### Expected Result
Only one large file is found (in the root of the product installation)

### Actual Result
Tycho 3.0.4 produces and packages two large files:
```
$ find org.basilevs.product/target/products/product/linux/gtk/x86_64 -size +900M  -ls             
25342712  2124272 -rw-r--r--    1 vasiligulevich   staff            1074069511 Jun 14 16:58 org.basilevs.product/target/products/product/linux/gtk/x86_64/p2/org.eclipse.equinox.p2.core/cache/binary/org.basilevs.rootfilesdemo_root_1.0.0.202306141248
25343330  2126960 -rw-r--r--    1 vasiligulevich   staff            1073741824 Jun 14 16:26 org.basilevs.product/target/products/product/linux/gtk/x86_64/large_root_file.bin
```

**Comment by basilevs** [comment]
When using [CBI signing plugin](https://github.com/eclipse-cbi/org.eclipse.cbi/tree/main/maven-plugins/eclipse-winsigner-plugin) to sign a custom Eclipse-based application, this cache still contains unsigned executable and triggers security vulnerabilities reports!

UPDATE 1: eclipse.exe in the cache of Eclipse Platform is signed
UPDATE 2: eclipse-inst.exe in the cache of Eclipse Installer does not pass signature verification.

**Comment by merks** [comment]
Yes, a product contains a p2 cache with the binary IUs used to install the product.

The SBOM generator I've been working on finds those:

https://download.eclipse.org/cbi/sbom/?file=https://download.eclipse.org/cbi/archive/sboms/cbi.p2repo.sbom-general-generator/66/eclipse-SDK-I20250828-0630-linux-gtk-x86_64-sbom.xml

Specifically these

<img width="1510" height="643" alt="Image" src="https://github.com/user-attachments/assets/b23fe7df-5c73-4421-b6db-1135aea88325" />

I believe this is simply how p2 behaves and how it records the binaries that it processes via touch point instructions, but I'm not 100% sure about that.  In any case, all products I see are like this.

---

You mention unsigned executables.  Signed executables are not supported on Linux as I understand it.

 On Windows there are these binaries in the cache:

<img width="983" height="150" alt="Image" src="https://github.com/user-attachments/assets/c160d937-71ef-4e0e-8dde-79cac2b57d26" />

They don't have an executable extension and in fact they are zip files. 

If I copy and rename them I can see  the contents of the zip:

<img width="1329" height="159" alt="Image" src="https://github.com/user-attachments/assets/e9cb0854-550b-4639-85d3-2296e7d3015d" />

If I extract that, the executable appears to have a digital signature:

<img width="499" height="282" alt="Image" src="https://github.com/user-attachments/assets/037428dd-119a-482d-bd38-4de19843cc70" />

Can you elaborate on which platforms you see problems and exactly which problems you see?

**Comment by merks** [comment]
I can’t reproduce that. It suggests your build is not signing the executable while the EPP build is.  Or maybe the branding iron is invaliding the signature. Does the eclipse installer raise the same red flags?

https://www.eclipse.org/downloads/download.php?file=/oomph/products/eclipse-jre-inst-win64.zip

**Comment by basilevs** [comment]
[Nothing in winsigner plugin indicates signing of archives](https://github.com/eclipse-cbi/org.eclipse.cbi/blob/67521c3d8ff4861a0ef66ab30bc2ff4117abd214/maven-plugins/eclipse-winsigner-plugin/src/main/java/org/eclipse/cbi/maven/plugins/winsigner/SignMojo.java#L318)
I suspect that EPP build deals with a presigned Windows executable, while custom product's executable is patched by Tycho (icon, publisher information, etc.) and therefore is a subject to resign.

**Comment by merks** [comment]
That’s why I ask about the installer because it’s a public example.

**Comment by basilevs** [comment]
I've confirmed that the signature is obsolete in a patched executable:

<img width="838" height="674" alt="Image" src="https://github.com/user-attachments/assets/09883d96-4634-41e7-83a6-14273dd475f4" />

The link to installer is incorrect, I'll download from the official location.
I can't run the scanner, but I can check if signature dates are matching.

**Comment by basilevs** [comment]
I've confirmed that digital signature for cached eclipse-inst.exe from [Eclipse Installer](https://www.eclipse.org/downloads/download.php?file=/oomph/epp/2025-06/R/eclipse-inst-jre-win-aarch64.exe) is invalid. @merks  thanks for the hint!

<img width="678" height="671" alt="Image" src="https://github.com/user-attachments/assets/0d40e231-6d05-4023-98fb-e68a730cb6ab" />

It is safe to assume that security scanners would flag Eclipse Installer.

**Comment by merks** [comment]
Yes, looking at these:

- https://www.eclipse.org/downloads/download.php?file=/oomph/products/eclipse-inst-jre-win64.exe
- https://www.eclipse.org/downloads/download.php?file=/oomph/products/eclipse-inst-jre-win64.zip

The overall self-extracting *.exe is signed.

<img width="404" height="692" alt="Image" src="https://github.com/user-attachments/assets/40789361-6400-4dc2-8c50-6bd84a0f41ca" />

The two *.exe in the zip are signed:

<img width="484" height="693" alt="Image" src="https://github.com/user-attachments/assets/7794a1e6-a09e-4c89-8366-dbf869469d25" />

<img width="500" height="620" alt="Image" src="https://github.com/user-attachments/assets/3a5e12c2-9bf9-4e64-a44f-f3ae4db99086" />

But the ones in the binaries are signed but with only one signature not two like the above:

<img width="537" height="646" alt="Image" src="https://github.com/user-attachments/assets/c12c3af6-bf75-4fe2-9343-a635ab5810a0" />

And that one invalid signature is likely the one that was applied before the branding iron, while the valid signature is only on the final executable surfaced in the product. 

---

For the installer product the winsigner is configured like this:

<img width="690" height="525" alt="Image" src="https://github.com/user-attachments/assets/e557860e-5b42-45f4-9de2-aaf0eacdc5cb" />

I'm not sure how to ensure these are signed:

<img width="718" height="706" alt="Image" src="https://github.com/user-attachments/assets/a7fae5ec-9a6d-4cbd-a7b1-1b379a2762c9" />

I believe they are packaged up from this location and that we should be signing them there before they are packaged up

<img width="387" height="370" alt="Image" src="https://github.com/user-attachments/assets/b6a4e591-a9ce-4b1f-aa5c-d8019c308a90" />

@laeubi 

I wonder what phase we'd need to do the signing in for the executables at that "interim" location?   (Probably then we'd not even need to do the signing at the final location because probably those files are copied to that location and would be signed before the copying.)

**Comment by sratz** [comment]
The problem actually is `target/extraArtifacts/` which is created during `tycho-p2-publisher-plugin:publish-products` and is the source for `target/repository/` and the materialized products.

What we have done in our product, is
- Move the `tycho-p2-publisher-plugin:publish-*` mojos from the `prepare-package` phase earlier to `generate-resources`
- In the `generate-resources` phase, then unzip the binaries in there
- In the `process-resources` phase, sign them
- In the `process-resources` phase, zip them again

Now, when materialize-products and or assemble-repository runs, they already see the signed files, they end up everywhere else (and the checksums are already created correctly, so no need to tidy up afterwards):

```xml
<project
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0">
  <modelVersion>4.0.0</modelVersion>
  <artifactId>product</artifactId>
  <version>3.55.0-SNAPSHOT</version>
  <name>${project.groupId} Product</name>
  <packaging>eclipse-repository</packaging>

  <parent>
    <artifactId>parent</artifactId>
    <groupId>com.foo.bar.installation</groupId>
    <version>3.55.0-SNAPSHOT</version>
  </parent>

  <build>

    <plugins>

      <!-- Move tycho-p2-publisher-plugin to an earlier phase by overriding the phase of their default executions,
           so that we can code-sign the exe file before it is being packaged.
           For some reason, this must be done outside of <profile>, has no effect otherwise. -->
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-publisher-plugin</artifactId>
        <version>${tycho-version}</version>
        <executions>
          <execution>
            <phase>generate-resources</phase>
            <id>default-publish-osgi-ee</id>
            <goals>
              <goal>publish-osgi-ee</goal>
            </goals>
          </execution>
          <execution>
            <phase>generate-resources</phase>
            <id>default-publish-products</id>
            <goals>
              <goal>publish-products</goal>
            </goals>
          </execution>
          <execution>
            <phase>generate-resources</phase>
            <id>default-publish-categories</id>
            <goals>
              <goal>publish-categories</goal>
            </goals>
          </execution>
          <execution>
            <phase>generate-resources</phase>
            <id>default-attach-artifacts</id>
            <goals>
              <goal>attach-artifacts</goal>
            </goals>
          </execution>
        </executions>
      </plugin>

      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-director-plugin</artifactId>
        <version>${tycho-version}</version>
        <configuration>
          <products>
            <product>
              <id>com.foo.bar.product.distribution</id>
            </product>
          </products>
        </configuration>
      </plugin>

      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-repository-plugin</artifactId>
        <version>${tycho-version}</version>
        <configuration>
          <includeAllDependencies>true</includeAllDependencies>
        </configuration>
      </plugin>

      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>target-platform-configuration</artifactId>
        <version>${tycho-version}</version>
        <configuration>
          <executionEnvironment>none</executionEnvironment>
        </configuration>
      </plugin>

    </plugins>
  </build>

  <profiles>

    <profile>
      <id>sign</id>
      <build>
        <plugins>
          <plugin>
            <artifactId>maven-antrun-plugin</artifactId>
            <executions>
              <execution>
                <id>unzip</id>
                <phase>generate-resources</phase>
                <configuration>
                  <target>
                    <unzip src="${project.build.directory}/extraArtifacts/com.foo.bar.product.distribution.executable.win32.win32.x86_64.zip" dest="${project.build.directory}/tosign/"/>
                  </target>
                </configuration>
                <goals>
                  <goal>run</goal>
                </goals>
              </execution>
              <execution>
                <id>zip</id>
                <phase>process-resources</phase>
                <configuration>
                  <target>
                    <!-- delete old zip first, zip file timestamps only have 2 seconds precision, so if signing is fast enough, things aren't updated otherwise... -->
                    <delete file="${project.build.directory}/extraArtifacts/com.foo.bar.product.distribution.executable.win32.win32.x86_64.zip"/>
                    <zip destfile="${project.build.directory}/extraArtifacts/com.foo.bar.product.distribution.executable.win32.win32.x86_64.zip" basedir="${project.build.directory}/tosign/"/>
                  </target>
                </configuration>
                <goals>
                  <goal>run</goal>
                </goals>
              </execution>
            </executions>
          </plugin>

          <plugin>
            <groupId>net.jsign</groupId>
            <artifactId>jsign-maven-plugin</artifactId>
            <version>7.2</version>
            <executions>
              <execution>
                <goals>
                  <goal>sign</goal>
                </goals>
                <phase>generate-resources</phase>
                <configuration>
                  <fileset>
                    <directory>${project.build.directory}/tosign</directory>
                  </fileset>
                  <name>...</name>
                  <url>...</url>
                  <tsaurl>...</tsaurl>
                  <keystore>...</keystore> <!-- jsign-maven-plugin requires this to be a file, not an URL! -->
                  <storepass>...</storepass>
                  <keypass>...</keypass>
                  <alias>...</alias>
                  <replace>true</replace>
                </configuration>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </profile>

  </profiles>

</project>
```

**Comment by basilevs** [comment]
@laeubi @merks note that for this particular issue, just wiping the cache would be enough. Will P2 be able to uninstall software if cache is wiped?

**Comment by merks** [comment]
Yes, that is the question I had at the top of my mind.  I suspect so, but I'm literally swamped with SimRel-related issues.  You could try it.

But the other important thing to keep in mind is that these binaries are also what's in the p2 repository, so if you update the product and it needs to update the native executables, this is where they will come from, so if they're not signed in the p2 repo that's going to remain a problem.

---

@sratz 

Thanks for the constructive suggestion.  It seems horribly complicated though for what is a general problem that anyone using branding will encounter. 

 @laeubi  

It would be nice if it were easier to hook signing into the build process without all this horrendous stuff!

**Comment by laeubi** [comment]
> It would be nice if it were easier to hook signing into the build process without all this horrendous stuff!

I have described the current process here and we maybe improve on that in the future:

- https://tycho.eclipseprojects.io/doc/main/SignProducts.html

for the sake of this issue, I think we should include an option in the product assembly mojo that simply deletes the cache folder.

**Comment by merks** [comment]
While deleting the cache folder helps to mask the problem, the underlying problem is that an actual update will update to an unsigned native executable because it's unsigned in the p2 repository.  But then masking does no harm and can make people complain less.  😀
